pkg/web/apps/apps_service.go

```diff
@@ -79,3 +81,103 @@ func (a *appsService) DisableAllAppVersionsByAppID(id string, message string) er
 
 	return a.repository.DisableAllAppVersionsByAppID(app.AppID, message)
 }
+
+// // InitClientApp updates the number of
+func (a *appsService) InitClientApp(deviceInfo *models.Device) (*models.Version, error) {
+	if _, err := a.repository.GetAppByAppID(deviceInfo.AppID); err != nil {
+		return nil, err
+	}
+
+	version, err := a.repository.GetVersionByAppIDAndVersion(deviceInfo.AppID, deviceInfo.Version)
+
+	// If any error other Not Found error occurred, return
+	if err != nil && err != models.ErrNotFound {
+		return nil, err
+	}
+
+	// If the version does not exist, create it
+	if err == models.ErrNotFound {
+		version = &models.Version{
+			ID:      uuid.New().String(),
+			Version: deviceInfo.Version,
+			AppID:   deviceInfo.AppID,
+		}
+	}
+
+	// Increment the number of app launches
+	version.NumOfAppLaunches++
+
+	// Update the existing version or create a new one
+	if err := a.repository.UpsertVersionWithAppLaunchesAndLastLaunched(version); err != nil {
+		return nil, err
+	}
+
+	device, err := a.repository.GetDeviceByDeviceIDAndAppID(deviceInfo.DeviceID, deviceInfo.AppID)
+
+	if err != nil {
+		// If we can't find the device by device ID and app ID
+		if err != models.ErrNotFound {
+			return nil, err
+		}
+		// Build a new device to save to the database
+		device = newDeviceFromDeviceAndVersion(*deviceInfo, *version)
+	}
+
+	if updateDeviceVersionID(device, version.ID) || updateDeviceDeviceVersion(device, deviceInfo.DeviceVersion) {
+
+		err := a.repository.InsertDeviceOrUpdateVersionID(*device)
+
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// clear these values before returning the data
+	version.LastLaunchedAt = ""
+	version.NumOfAppLaunches = 0
+
+	return version, nil
+}
+
+// If the Device.VersionID property is different to @var versionID,
+// update it to match, returning a bool to indicate if it was updated
+func updateDeviceVersionID(device *models.Device, versionID string) bool {
+	var isUpdated bool
+
+	if device.VersionID != versionID {
+		device.VersionID = versionID
+
+		isUpdated = true
+	}
+
+	return isUpdated
+}
+
+// If the Device.DeviceVersion property is different to @var deviceVersion,
+// update it to match, returning a bool to indicate if it was updated
+func updateDeviceDeviceVersion(device *models.Device, deviceVersion string) bool {
```

Comment:
Update device device?

Reply:
Why it is required?

Reply:
Update the DeviceVersion property of the Device model.

Reply:
IHMO it is not required. 
The device strtuct should have constructors. 
In this way you can have: `model.Device.New(deviceInfo, version)`  
which will be something as 

```go
func New(deviceInfo ..., version ..) modelDevice{
      return &device{
         params
       }
}
```

Also, after you have the device you can do things like `device.versionInfoN = version.InfoN`

Is it make sense?

Reply:
I'd like to recommend you check how to create constructors. See it [here](https://golang.org/doc/effective_go.html#composite_literals) for example.

Reply:
@camilamacedo86 this is not a constructor function. This is updating a specific property on the device model based on a condition.

Reply:
I still not getting the need for this impl. What is the need and when it should be done? Why you need do this implementation? To solve what scenario?

Reply:
Because if the condition is met then an update takes place. See [here](https://github.com/aerogear/mobile-security-service/blob/3fa0f2a5693f3b79102015088846dd1d5a1678f9/pkg/web/apps/apps_service.go#L157-L158)

Reply:
It still not making sense. 


Why you need a func updateDeviceDeviceVersion + the ifs and code implementations in the service when it is simple as follows.? 

> IF NOT exist a device with ANY version of the same app
> *  create the device
> ELSE
> * update the version number in the device

The service implementation made here is too unnecessary complex, as you check in the [suggestion](https://github.com/aerogear/mobile-security-service/pull/76#discussion_r262879241) which is a pseudo code it can be really more simple, clean, and with high cohesion, and more understandable. If get the text and replace it for the code you will have the service implemented. 

Note that the solution implemented here as I described above will definitely face concurrency issues and in this way will not attend the no-functional requirements of this system. 

c/c @davidffrench